### PR TITLE
*: support in subquery unfolding

### PR DESCRIPTION
**Pull request by hanfei1991**
add a flag to unfold in subquery.
@shenli @coocood @zimulala @lamxTyler @winoros PTAL

### CODE REVIEW

File: plan/expression_rewriter.go [hunk 1]
```diff
@@ -487,6 +487,37 @@ func (er *expressionRewriter) handleInSubquery(v *ast.PatternInExpr) (ast.Node,
 		er.err = ErrOperandColumns.GenByArgs(lLen)
 		return v, true
 	}
+	if lLen == 1 && asScalar && er.ctx.GetSessionVars().AllowInSubqueryUnFolding {
```
Comment (shenli) [inline plan/expression_rewriter.go]: Add some comments for this code block.

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -617,7 +648,9 @@ func (er *expressionRewriter) Leave(inNode ast.Node) (retNode ast.Node, ok bool)
 	case *ast.RowExpr:
 		er.rowToScalarFunc(v)
 	case *ast.PatternInExpr:
-		er.inToExpression(v)
+		if v.Sel == nil {
```
Comment (shenli) [inline plan/expression_rewriter.go]: Why should we check this?

File: plan/logical_plan_builder.go [hunk 3]
```diff
@@ -54,7 +54,6 @@ func (b *planBuilder) buildAggregation(p LogicalPlan, aggFuncList []*ast.Aggrega
 		baseLogicalPlan: newBaseLogicalPlan(Agg, b.allocator)}
 	agg.self = agg
 	agg.initIDAndContext(b.ctx)
-	addChild(agg, p)
```
Comment (shenli) [inline plan/logical_plan_builder.go]: Why move this down?

Comment (hanfei1991) [inline plan/logical_plan_builder.go]: The p may be changed after expr rewriting.

File: plan/expression_rewriter.go (hunk 2 above)
Comment (hanfei1991) [inline plan/expression_rewriter.go]: if v.Sel is not nil, it has been proccessed. I move this check out of the `inToExpression`.

File: executor/executor.go [hunk 4]
```diff
@@ -349,24 +349,26 @@ func init() {
 	// While doing optimization in the plan package, we need to execute uncorrelated subquery,
 	// but the plan package cannot import the executor package because of the dependency cycle.
 	// So we assign a function implemented in the executor package to the plan package to avoid the dependency cycle.
-	plan.EvalSubquery = func(p plan.PhysicalPlan, is infoschema.InfoSchema, ctx context.Context) (d []types.Datum, err error) {
+	plan.EvalSubquery = func(p plan.PhysicalPlan, is infoschema.InfoSchema, ctx context.Context) (rows [][]types.Datum, err error) {
```
Comment (hanfei1991) [inline executor/executor.go]: For Exist subquery, we append a Limit 1 to the end of the plan.

**Comment by hanfei1991** [comment]
@shenli PTAL

**Comment by shenli** [comment]
LGTM
@coocood PTAL

**Comment by hanfei1991** [comment]
@shenli PTAL again

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood @zimulala PTAL

File: plan/expression_rewriter.go [hunk 5]
```diff
@@ -487,6 +487,51 @@ func (er *expressionRewriter) handleInSubquery(v *ast.PatternInExpr) (ast.Node,
...
+				RetType: np.Schema().Columns[0].GetType(),
+			}
+			er.ctxStack = append(er.ctxStack, con)
+		}
+		listLen := len(rows)
+		if listLen == 0 {
+			er.ctxStack[len(er.ctxStack)-1] = &expression.Constant{
+				Value:   types.NewDatum(false),
+				RetType: types.NewFieldType(mysql.TypeTiny),
+			}
```
Comment (coocood) [inline plan/expression_rewriter.go]: Why do not need to add Selection here if asScalar is false?

Comment (hanfei1991) [inline plan/expression_rewriter.go]: If it is false, we should return a scalar function

**Comment by hanfei1991** [comment]
@coocood PTAL

File: executor/aggregate_test.go [hunk 6]
```diff
@@ -272,6 +272,19 @@ func (s *testSuite) TestAggregation(c *C) {
...
+	tk.MustExec("create table t1 (c1 int)")
+	tk.MustExec("create table t2 (c1 int)")
+	tk.MustExec("insert into t1 values(3), (2)")
+	tk.MustExec("insert into t2 values(1), (2)")
+	tk.MustExec("set @@session.tidb_opt_insubquery_unfold = 1")
+	result = tk.MustQuery("select sum(c1 in (select * from t2)) from t1")
+	result.Check(testkit.Rows("1"))
+	tk.MustExec("set @@session.tidb_opt_insubquery_unfold = 0")
+	result = tk.MustQuery("select sum(c1 in (select * from t2)) from t1")
+	result.Check(testkit.Rows("1"))
```
Comment (coocood) [inline executor/aggregate_test.go]: Please add test for unfolding a in subquery in where.

Comment (coocood) [inline executor/aggregate_test.go]: cover the empty and non-empty result.

**Comment by coocood** [comment]
LGTM

**Pull request merged**